### PR TITLE
Admin Generator (Future): Allow adding custom grid actions

### PR DESCRIPTION
**Pull request by jamesricky**
---

## PR Checklist

-   [x] Verify if the change requires a changeset. See [CONTRIBUTING.md](https://github.com/vivid-planet/comet/blob/HEAD/CONTRIBUTING.md)
-   [x] Link to the respective task if one exists: SVK-211
-   [x] Provide screenshots/screencasts if the change contains visual changes

https://github.com/vivid-planet/comet/assets/6264317/e50be8d7-2a5c-4c78-bee8-b90639783e49

### CODE REVIEW

File: packages/admin/cms-admin/src/generator/future/generator.ts [hunk 1]
```diff
@@ -56,6 +56,10 @@ export type GridConfig<T extends { __typename?: string }> = {
     delete?: boolean;
     copyPaste?: boolean;
     readOnly?: boolean;
+    actions?: {
```
Comment (nsams) [inline packages/admin/cms-admin/src/generator/future/generator.ts:59]: for actions (plural) I would expect I can specify an array for multiple actions

File: packages/admin/cms-admin/src/generator/future/generator.ts [hunk 2]
```diff
@@ -56,6 +56,10 @@ export type GridConfig<T extends { __typename?: string }> = {
...
+        columnWidth?: number;
+        componentImport?: ImportReference;
```
Comment (nsams) [inline packages/admin/cms-admin/src/generator/future/generator.ts:61]: the ImportReference for block/validate is not named with ...Import, I would also not do that here

File: packages/admin/cms-admin/src/generator/future/generator.ts (hunk 1 above)
Comment (jamesricky) [reply packages/admin/cms-admin/src/generator/future/generator.ts:59]: You can specify a component, that has multiple actions 😄 

Should we change this to two settings?

- `actionsComponent?: ImportReference`
- `actionsWidth?: number`

**Comment by nsams** [comment]
oh, and have you considered https://mui.com/x/react-data-grid/column-definition/#special-properties `type: "actions"` insted of our actions component?

**Comment by jamesricky** [comment]
> oh, and have you considered https://mui.com/x/react-data-grid/column-definition/#special-properties `type: "actions"` insted of our actions component?

No, good point. 

It seems like it could make sense to use this internally, but then we'd need to allow an array of component-imports as `getActions()` expects an array of individual action-components. 
The current approach (in this PR) only needs a single component-import, as that can be a component with multiple actions. 
A setting for the width would also still be needed.

File: packages/admin/cms-admin/src/generator/future/generator.ts (hunk 1 above)
Comment (jamesricky) [reply packages/admin/cms-admin/src/generator/future/generator.ts:60]: @nsams should I do it the way I suggested?

**Comment by nsams** [comment]
> but then we'd need to allow an array of component-imports as getActions() expects an array of individual action-components.

Yes, why not, it's always a good idea to stay close to MUI api

File: demo/admin/src/products/future/ProductsGrid.cometGen.ts [hunk 3]
```diff
@@ -14,4 +14,8 @@ export const ProductsGrid: GridConfig<GQLProduct> = {
         { type: "date", name: "availableSince", width: 140 },
         { type: "dateTime", name: "createdAt", width: 170 },
     ],
+    actions: {
+        columnWidth: 116,
```
Comment (johnnyomair) [inline demo/admin/src/products/future/ProductsGrid.cometGen.ts:18]: Why is this needed?

File: demo/admin/src/products/future/ProductsGrid.cometGen.ts [hunk 4]
```diff
@@ -14,4 +14,8 @@ export const ProductsGrid: GridConfig<GQLProduct> = {
         { type: "date", name: "availableSince", width: 140 },
         { type: "dateTime", name: "createdAt", width: 170 },
     ],
+    actions: {
```
Comment (johnnyomair) [inline demo/admin/src/products/future/ProductsGrid.cometGen.ts:17]: Have you considered having `type: "actions"` in the `columns`? This would allow overriding the default actions and placing the actions anywhere in the grid. Do we even want that?

**Comment by jamesricky** [comment]
> > but then we'd need to allow an array of component-imports as getActions() expects an array of individual action-components.
> 
> Yes, why not, it's always a good idea to stay close to MUI api

If we would simply import and pass in a react component, like the MUI api expects I would agree. 
In the generator however we need configuration for every component to import. 
Since the configuration is just strings (`name`, `import`), the IDE doesn't know to check if the file or the component exists. 
I think it would be good to avoid this kind of error-prone/non-autocompleted configuration as much as possible, that's why I'd prefer a single import with one component that has multiple IconButtons.

File: packages/admin/cms-admin/src/generator/future/generator.ts (hunk 2 above)
Comment (jamesricky) [reply packages/admin/cms-admin/src/generator/future/generator.ts:62]: > Can the import really be optional?

It's possible that someone would like to adjust the column-width independent of wanting to add an additional component. 
Although, when there is no additional component the width should probably be calculated, depending on how many actions are shown (Edit, Context-Menu, nothing) that is currently not the case however.

File: demo/admin/src/products/future/ProductsGrid.cometGen.ts (hunk 4 above)
Comment (nsams) [reply demo/admin/src/products/future/ProductsGrid.cometGen.ts:17]: I also think this would be natural, as it's a column after all

File: demo/admin/src/products/future/ProductsGrid.cometGen.ts (hunk 3 above)
Comment (jamesricky) [reply demo/admin/src/products/future/ProductsGrid.cometGen.ts:18]: DataGrid columns do not scale automatically, depending on their content. 
We need to set the width manually depending on the number of `IconButton` components shown to prevent them from being cut off.

**Comment by jamesricky** [comment]
Setting this to "Draft" for now, until https://github.com/vivid-planet/comet/pull/2044 is approved.

File: demo/admin/src/products/future/ProductsGrid.cometGen.ts (hunk 3 above)
Comment (nsams) [reply demo/admin/src/products/future/ProductsGrid.cometGen.ts:18]: how does mui `type: "actions"` solve this width issue? Do you also have to specify it manually?

Comment (jamesricky) [reply demo/admin/src/products/future/ProductsGrid.cometGen.ts:18]: It doesn't, you also have to specify it manually.

File: packages/admin/cms-admin/src/generator/future/generateGrid.ts [hunk 5]
```diff
@@ -155,6 +155,11 @@ export function generateGrid(
 
     const showActionsColumn = allowCopyPaste || allowEditing || allowDeleting;
 
+    const actionsColumnProps = ['field: "actions"', 'headerName: ""', "sortable: false", "filterable: false", 'type: "actions"', 'align: "right"'];
+    if (typeof config.actions?.columnWidth !== "undefined") {
+        actionsColumnProps.push(`width: ${config.actions.columnWidth}`);
+    }
```
[the remark this answers is not in the document]
Comment (jamesricky) [reply packages/admin/cms-admin/src/generator/future/generateGrid.ts:141]: So I can add the if without adding another if inside the already confusingly large conditional template-string 😄 
And I didn't want to set `width={undefined}` as the fallback if no width is set.

File: packages/admin/cms-admin/src/generator/future/generateGrid.ts [hunk 6]
```diff
@@ -487,23 +493,18 @@ export function generateGrid(
                 ${
                     showActionsColumn
                         ? `{
-                        field: "actions",
-                        headerName: "",
-                        sortable: false,
-                        filterable: false,
-                        type: "actions",
-                        align: "right",
+                        ${actionsColumnProps.join(",\n")},
```
Comment (nsams) [inline packages/admin/cms-admin/src/generator/future/generateGrid.ts:496]: you might consider using `tsCodeRecordToString({
    field: "\"actions\"",
    headerName: "\"\"",
    "sortable: false",
//   ....
})`

**Review by nsams — APPROVED** [review]

File: packages/admin/cms-admin/src/generator/future/generateGrid.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jamesricky) [reply packages/admin/cms-admin/src/generator/future/generateGrid.ts]: I changed this, as I now use `tsCodeRecordToString()`

File: packages/admin/cms-admin/src/generator/future/generateGrid.ts [hunk 7]
```diff
@@ -20,11 +20,17 @@ import { generateImportsCode, Imports } from "./utils/generateImportsCode";
 
 type TsCodeRecordToStringObject = Record<string, string | number | undefined>;
 
-function tsCodeRecordToString(object: TsCodeRecordToStringObject) {
-    return `{${Object.entries(object)
+function tsCodeRecordToString(object: TsCodeRecordToStringObject, removeBrackets = false) {
```
Comment (nsams) [inline packages/admin/cms-admin/src/generator/future/generateGrid.ts:23]: why this argument?

Comment (jamesricky) [reply packages/admin/cms-admin/src/generator/future/generateGrid.ts]: Not all keys were defined inside the `tsCodeRecordToString` so the `{}` needed to be defined outside. 
I just moved everything inside `tsCodeRecordToString` now, so it's no longer needed.

**Review by nsams — APPROVED** [review]

File: demo/admin/src/products/future/ProductsGrid.cometGen.ts (hunk 4 above)
Comment (johnnyomair) [reply demo/admin/src/products/future/ProductsGrid.cometGen.ts]: @jamesricky what about this suggestion?

File: demo/admin/src/products/future/ProductsGrid.cometGen.ts [hunk 8]
```diff
@@ -15,4 +15,6 @@ export const ProductsGrid: GridConfig<GQLProduct> = {
         { type: "date", name: "availableSince", width: 140 },
         { type: "dateTime", name: "createdAt", width: 170 },
     ],
+    actionsWidth: 116,
+    actionsComponent: { name: "ProductsGridPreviewAction", import: "../../ProductsGridPreviewAction" },
```
Comment (johnnyomair) [inline demo/admin/src/products/future/ProductsGrid.cometGen.ts:19]: To me this sounds like I can replace the actions component altogether. Maybe add an `additional` prefix?

File: demo/admin/src/products/future/ProductsGrid.cometGen.ts (hunk 4 above)
Comment (jamesricky) [reply demo/admin/src/products/future/ProductsGrid.cometGen.ts]: I agree, this api is better. 
Changed it in c972a96b68cbb396facc5e795d16c8ba3bd2ca69

File: packages/admin/cms-admin/src/generator/future/generator.ts [hunk 9]
```diff
@@ -81,6 +84,8 @@ export type GridConfig<T extends { __typename?: string }> = {
     toolbar?: boolean;
     toolbarActionProp?: boolean;
     rowActionProp?: boolean;
+    actionsComponent?: ImportReference;
+    actionsWidth?: number;
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/future/generator.ts:88]: This isn't needed anymore, right?

**Review by johnnyomair — APPROVED** [review]
LGTM. Please wait for @nsams review though.

File: demo/admin/src/products/ProductsGridPreviewAction.tsx [hunk 10]
```diff
@@ -1,15 +1,17 @@
...
 import { Dialog, DialogContent, DialogTitle, IconButton, Typography } from "@mui/material";
+import { GridCellParams } from "@mui/x-data-grid-pro";
+import { GQLProductsGridFutureFragment } from "@src/products/future/generated/ProductsGrid.generated";
 import { GQLProductsListManualFragment } from "@src/products/ProductsGrid.generated";
 import React from "react";
 import { FormattedMessage } from "react-intl";
 
 type Props = {
-    product: GQLProductsListManualFragment;
+    params: GridCellParams<unknown, GQLProductsListManualFragment | GQLProductsGridFutureFragment>;
```
Comment (nsams) [inline demo/admin/src/products/ProductsGridPreviewAction.tsx:11]: a prop named `params` is somehow strange.

- maybe name it `cellParams`?
- or skip this level: `type Props = GridCellParams<...>`?

Comment (jamesricky) [reply demo/admin/src/products/ProductsGridPreviewAction.tsx]: Good point, changed it to `type Props = GridCellParams<...>`

**Review by nsams — APPROVED** [review]
